### PR TITLE
feat/fix : Search feature added, Error handling enhanced, All Accounts functions added to API Service, Color Theme fix.

### PR DESCRIPTION
**Pull request by AbdulMalikDev**
# Changes Overview
- Error messages while logging in are more descriptive
  - **Expected** error messages
      1. Condition : **User not connected to internet**
          - Error Message: _Network Connection Error_
     2. Condition : **Username Empty**
        - Error Message: _Username cannot be empty!_
     3. Condition : **Password Empty**
        - Error Message: _Password cannot be empty!_
     4. Condition : **Valid URL, Username and Password but attempt to connect to Server Failed**
        - Error Message: _Username or Password incorrect_
     > ⬆️ Essentially how this is checked is by making a request to the server to get the disk space available.
     > If the response is valid and diskspace is received the accound credentials are verified.
     5. Condition : **Valid URL, Username and Password, attempt to connect to Server without error but response code != 200**
        - Error Message: _Something\'s Wrong_
<br/>

- Application Version is now fetched (async) and stored while **loading screen** is shown to user. Before the PR this used to load every time the user opened the drawer resulting in a 1 second delay.

- Disk Explorer Screen now has Search Functionality as well as Sorting Functionality.

- History Screen also now has Search Functionality as well as Sorting Functionality.

- Dark Mode settting now locally stored. Tested by building apk, installing in device.

- Following functions added in the API Service to support the "All Accounts" feature.
  - getAllAccountsHistory
  - updateAllAccountsHistory
  - getAllAccountsDiskFiles
  - loadAllAccountsRSS
  - getAllAccountsRSSFilters

- Selected Color of Filters and Labels enhanced in Home Screen Drawer

Light Mode               |  Dark Mode               
:-------------------------:|:-------------------------:
||

- Unit Testing structure added complying with the stacked architecture.

### CODE REVIEW

File: lib/ui/views/login/login_viewmodel.dart [hunk 1]
```diff
@@ -41,11 +41,10 @@ class LoginViewModel extends BaseViewModel {
...
     setBusy(true);
-    if (username.contains(" ") ||
-        password!.contains(" ") ||
-        username.isEmpty ||
-        password.isEmpty) {
-      Fluttertoast.showToast(msg: 'Invalid username or password');
+    if (username.trim().isEmpty) {
+      Fluttertoast.showToast(msg: 'Username cannot be empty!');
+    } else if(password!.trim().isEmpty){
+      Fluttertoast.showToast(msg: 'Password cannot be empty!');
```
Comment (rj-since-2000) [inline lib/ui/views/login/login_viewmodel.dart]: What's the need to check for trim and empty strings if it already has been checked above?

You may want to create separate checks for username and password and then display "Username cannot be empty" and "Password cannot be empty" respectively.

Also the error message "Invalid username or password" is for validating the credentials from the server and not for the local check for empty credentials.

File: lib/services/state_services/disk_file_service.dart [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      displayList = displayList
+          .where((element) =>
+              element.name!.toLowerCase().contains(searchText.toLowerCase()))
+          .toList();
+    }
+    diskFileDisplayList.value = displayList;
+    // ignore: invalid_use_of_visible_for_testing_member
+    // ignore: invalid_use_of_protected_member
+    diskFileDisplayList.notifyListeners();
+  }
```
Comment (rj-since-2000) [inline lib/services/state_services/disk_file_service.dart:53]: How will the displayList of Disk Files reset to the original list ?

**Review by rj-since-2000 — CHANGES_REQUESTED** [review]
Make the necessary changes in your new PR

File: lib/services/state_services/history_service.dart [hunk 3]
```diff
@@ -40,11 +47,53 @@ class HistoryService extends ChangeNotifier {
...
+
+    if (searchText != null) {
+      //Searching : showing list on basis of searched text
+      displayList = displayList
+          .where((element) =>
+              element.name.toLowerCase().contains(searchText.toLowerCase()))
+          .toList();
+    }
+    _torrentsHistoryDisplayList.value = displayList;
     _torrentsHistoryDisplayList.notifyListeners();
```
[the remark this answers is not in the document]
Comment (AbdulMalikDev) [reply lib/services/state_services/history_service.dart:65]: There are two ways for this to happen.
1. User presses Filter option and presses the "All" button
2. User chooses the Sort.None option

**Pull request closed without merging**